src/EFCore.Relational/Query/SqlExpressions/SelectExpression.Helper.cs

```
@@ -1008,6 +1009,29 @@ private sealed class CloningExpressionVisitor : ExpressionVisitor
                 return newTpcTable;
             }
 
+            if (expression is TableValuedFunctionExpression tableValuedFunctionExpression)
```

Maybe make TableValuedFunctionExpression implement IClonableTableExpressionBase and move the code into there? Could even do the same for TpcTablesExpression above (and maybe even SelectExpression)?

---

You're right, actual cloning can't really be moved out of the visitor (at least not without a deeper rethink of the design).

Re failing faster... Is the problem specific to TableExpressionBase for some reason, i.e. do we specifically need to implement custom logic for those as opposed to other expression types? In general, expression types which are immutable should read to be cloned right? The big exception is obviously SelectExpression, but do TableExpressionBase also pose a problem?